tests/integration/security/ca_custom_root/ecc/trust_domain_validation_test.go

```diff
@@ -0,0 +1,218 @@
+// +build integ
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecc
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	client2 "istio.io/istio/pkg/test/echo/client"
+	"istio.io/istio/pkg/test/echo/common/scheme"
+	epb "istio.io/istio/pkg/test/echo/proto"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+const (
+	httpPlaintext = "http-plaintext"
+	httpMTLS      = "http-mtls"
+	tcpPlaintext  = "tcp-plaintext"
+	tcpMTLS       = "tcp-mtls"
+	passThrough   = "tcp-mtls-pass-through"
+
+	// policy to enable mTLS in client and server:
+	// ports with plaintext: 8090 (http) and 8092 (tcp)
+	// ports with mTLS: 8091 (http), 8093 (tcp) and 9000 (tcp passthrough).
+	policy = `
+apiVersion: "security.istio.io/v1beta1"
+kind: "PeerAuthentication"
+metadata:
+  name: "mtls"
+spec:
+  selector:
+    matchLabels:
+      app: server
+  mtls:
+    mode: STRICT
+  portLevelMtls:
+    8090:
+      mode: DISABLE
+    8092:
+      mode: DISABLE
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: server
+spec:
+  host: server.%s.svc.cluster.local
+  trafficPolicy:
+    tls:
+      mode: ISTIO_MUTUAL
+    portLevelSettings:
+    - port:
+        number: 8090
+      tls:
+        mode: DISABLE
+    - port:
+        number: 8092
+      tls:
+        mode: DISABLE
+`
+)
+
+// TestTrustDomainValidation tests the trust domain validation when mTLS is enabled.
+// The trust domain validation should reject a request if it's not from the trust domains configured in the mesh config.
+// The test uses naked client (no sidecar) with custom certificates of different trust domains and covers the following:
+// - plaintext requests are not affected
+// - same trust domain (cluster.local) and aliases (trust-domain-foo and trust-domain-bar)
+// - works for both HTTP and TCP protocol
+// - works for pass through filter chains
+func TestTrustDomainValidation(t *testing.T) {
```

Comment:
can we reuse the same test code? i haven't looked at the other test but I imagine this is copy+pasted?

Reply:
That's the intent. There's much cleanup to do. For now I was trying to make sure getting the env set correctly was working. It looks like I may have to put it in its own top-level test. Is that correct? I'm not sure why there are test failures. I would have thought putting it in its own package would suffice.